src/chromatix/functional/propagation.py

```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
     delay = jnp.sqrt(jnp.abs(kernel))
     delay = jnp.where(kernel >= 0, delay, 1j * delay)  # keep evanescent modes
     phase = 2 * jnp.pi * (z * n / field.spectrum) * delay
-    return jnp.fft.ifftshift(jnp.exp(1j * phase), axes=field.spatial_dims)
+    kernel_field = jnp.exp(1j * phase)
+
+    if bandlimit:
+        Sy, Sx = (1 / field.dk).squeeze()   # spatial dimension in microns
+        y0, x0 = (kykx / field.dk).squeeze()  # spatial shift in microns, TODO check
+        z0 = z.squeeze()  # propagation distance in microns
+        wv = field.spectrum.squeeze()  # wavelength in microns
+
+        dfX = 1.0 / Sx
+        dfY = 1.0 / Sy
+        N_y, N_x = field.spatial_shape
+        fX = np.linspace(-N_x // 2, N_x // 2 - 1, num=N_x)[np.newaxis, :] * dfX
+        fY = np.linspace(-N_y // 2, N_y // 2 - 1, num=N_y)[:, np.newaxis] * dfY
+
+        # Table 1 of "Shifted angular spectrum method for off-axis numerical 
+        # propagation" (2010) by Matsushima
+        du = 1 / (2 * Sx)
+        u_limit_p = ((x0 + 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        u_limit_n = ((x0 - 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sx < x0:
+            u0 = (u_limit_p + u_limit_n) / 2
+            u_width = u_limit_p - u_limit_n
+        elif x0 <= -Sx:
+            u0 = -(u_limit_p + u_limit_n) / 2
+            u_width = u_limit_n - u_limit_p
+        else:
+            u0 = (u_limit_p - u_limit_n) / 2
+            u_width = u_limit_p + u_limit_n
+
+        dv = 1 / (2 * Sy)
+        v_limit_p = ((y0 + 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        v_limit_n = ((y0 - 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sy < y0:
+            v0 = (v_limit_p + v_limit_n) / 2
+            v_width = v_limit_p - v_limit_n
+        elif y0 <= -Sy:
+            v0 = -(v_limit_p + v_limit_n) / 2
+            v_width = v_limit_n - v_limit_p
+        else:
+            v0 = (v_limit_p - v_limit_n) / 2
+            v_width = v_limit_p + v_limit_n
+
+        fx_max = u_width / 2
+        fy_max = v_width / 2
+
+        # bandlimit
+        H_filter = (np.abs(fX - u0) <= fx_max) * (np.abs(fY - v0) < fy_max)
+
+        # to jax
```

Comment:
probably a better way to do above with Jax?

Reply:
This is making a rectangle mask in the frequency grid? If so we shouldn't need `jnp.where` and this should just work because it's a binary mask and not variably sized (assuming the lines above can have their conditionals removed).

Reply:
Yes it's making a rectangular mask as shown in [Fig 5](https://opg.optica.org/getimagev2.cfm?img=7dKLEDezXUrkoUkzVOSYXrMOHiBaTdFKRs1SIJ54Pfo%3D&size=full&uri=oe-18-17-18453-g005) of the off-axis paper. I'll see if there's a way to remove the conditions before.

Reply:
done!